Add default filters

saroad2 opened the pull request:
Fixes #131 

Now, *pytest-cases* have a bunch of default filter functions to use in order to filter cases. 
Those filters can use "and" "or" and "invert" operations in order to create various complex filters.

smarie commented:
Thanks @saroad2 for this nice PR. In order to ease the review process would you mind creating a very small PR with just the migration of `_CaseInfo` and `_tags_matches_query` to `pytest_cases/case_info.py` first ? Or (faster for your) to revert this part, and propose the file migration later ? That way we'll be able to directly see what has changed in those symbols.

The test is very explicative, thanks.

I'm wondering how this can be documented in the file `docs/index.md` in addition to the other options. Can you propose some modification in the PR to this section: https://smarie.github.io/python-pytest-cases/#filters-and-tags  ? Also finally, since this introduces some API change, you'll have to modify `docs/api_reference.md` too. Sorry, this is not generated automatically as I find this almost as fast to copy this manually, than to try to find a no-too-bad configuration of sphinx :D

Finally note that #147 has also suggested some evolution of the regex engine (in dedicated ticket #153). I do not think that these are contradictory, that will make one more way to perform the same thing, and developers will choose whatever style they prefer.

saroad2 commented:
Hey @smarie ,

As you requested, I moved the `_CaseInfo` module change to PR #185 .

Also, I added the changes to `docs/index.md` in this PR.

Unfortunately, I wasn't sure where to add the new API into `docs/api_reference.md`. I would appreciate it if you would make that change instead.

Let me know if you want me to change anything else!

smarie commented:
It looks perfect. Ok I'll make the change to api reference myself once merged. Let's merge #185  first

saroad2 commented:
Hey @smarie , I merged master into this PR and updated `CaseInfo` after renaming.

Just add the new API to the documentation and this PR can be merged :)

Would you mind release a new version soon so I could use those new filters?

smarie commented on pytest_cases/filters.py:
```diff
@@ -0,0 +1,57 @@
+from .case_info import CaseInfo
+
+
+class CaseFilter:
```
py2 compat just in case

```suggestion
class CaseFilter(object):
```

smarie commented on pytest_cases/filters.py:
```diff
@@ -0,0 +1,57 @@
+        )
+
+
+def has_tag(tag_name):
```
```suggestion
def has_tag(tag_name):
    """Select cases that have the tag `tag_name`. See `@case(tags=...)` to add tags to a case."""
```

smarie commented on pytest_cases/tests/pytest_extension/parametrize_plus/test_filter.py:
```diff
@@ -0,0 +1,67 @@
+
+
+@parametrize_with_cases(
+    argnames="value", cases=".", filter=filters.has_prefix("case_t")
```
See above: if you had written "data_t" here you would not be able to retrieve anythin, because `prefix="case_"` and is applied before filters are executed. We should change this to `has_id_prefix` as suggested above.

smarie commented on pytest_cases/tests/pytest_extension/parametrize_plus/test_filter.py:
```diff
@@ -0,0 +1,67 @@
+
+
+@parametrize_with_cases(
+    argnames="value", cases=".", filter=filters.has_suffix("e")
```
Same comment as for prefix: rename id suffix

smarie commented on pytest_cases/filters.py:
```diff
@@ -0,0 +1,58 @@
+    return CaseFilter(lambda case: tag_name in CaseInfo.get_from(case).tags)
+
+
+def has_prefix(prefix):
+    return CaseFilter(lambda case: case.__name__.startswith(prefix))
```
The issue here is that `parametrize_with_cases` already has a `prefix` argument, and this is **always** used. Besides, it is used before the filters are called. Finally it does not take into account overridden `id`s.

So I suggest that we modify this so that it acts on the case id - that will complement nicely the `prefix` argument in order to work on the case `id`, not on the case function name.

```suggestion
def has_id_prefix(prefix):
    """Select cases that have a case id prefix `prefix`. Note that this is not the prefix of the case function, just the id (once the case prefix has been removed)"""
    return CaseFilter(lambda case: CaseInfo.get_from(case).id.startswith(prefix))
```

Note that tests should be updated accordingly, and the "suffix" filter too.

smarie commented on pytest_cases/filters.py:
```diff
@@ -0,0 +1,58 @@
+    return CaseFilter(lambda case: case.__name__.startswith(prefix))
+
+
+def has_suffix(suffix):
+    return CaseFilter(lambda case: case.__name__.endswith(suffix))
```
```suggestion
def has_suffix(suffix):
    return CaseFilter(lambda case: case.__name__.endswith(suffix))
```
def has_suffix(suffix):
    """Select cases that have a case id suffix `suffix`. 
    Note that this is not the suffix of the whole case function name, but the case id, possibly overridden with `@case(id=)`
    """
    return CaseFilter(lambda case: CaseInfo.get_from(case).id.startswith(prefix))

smarie commented on pytest_cases/filters.py:
```diff
@@ -0,0 +1,58 @@
+    return CaseFilter(lambda case: case.__name__.endswith(suffix))
+
+
+def match_regex(regex):
+    return CaseFilter(lambda case: re.match(regex, case.__name__))
```
```suggestion
def id_match_regex(regex):
    """Select cases that have a case id matching `regex`. 
    Note that this is not a match of the whole case function name, but the case id, possibly overridden with `@case(id=)`
    """
    return CaseFilter(lambda case: re.match(regex, CaseInfo.get_from(case).id))
```

smarie commented:
Thanks @saroad2 , I am deeply sorry to not have spotted this earlier but there are a few issues with the filters you propose: indeed the `prefix` argument or `parametrize_with_cases` is always applied at cases collection step, before filters are even applied. So I proposed a few changes where the filters apply always on the case id, not the case function name in order to not bring confusion/impossible filtering.

I am sorry that this takes longer than expected, I understand that this "back and forth" probably extremely frustrating for you... 

Unfortunately I received a very bad news today: my travis.com credits are now zero, I made a request for OSS credits but no build (no release) can be done until further notice.

I was planning to migrate all my travis script to nox, so this may force me to do it earlier than expected... Then once it is nox-ified I'll be independent on the build system and therefore be able to integrate in github actions, migrate back to travis, move to circleci... more freely as needed.

saroad2 commented:
@smarie , you shouldn't be sorry at all! we are al busy people and we have previous commitments, me too. This is an open source project, and as so everyone knows that it takes time for everything to get in motion. We are not in a hurry, so take your time!

As for the fixes you suggested, I totally agree. You don't need to apologise. I'll shortly make a commit with those changes.

As for Travis, in my experience, I had a lot of troubles with 3rd party CI/CD tools such as CircleCI/Travis/etc.
My suggestion is to migrate everything to Github Actions workflow and avoid using those 3rd parties tools.
Once I made that migration myself on my projects, I started to have a much better experience in my CI/CD process.

smarie commented on pytest_cases/tests/pytest_extension/parametrize_plus/test_filter.py:
```diff
@@ -0,0 +1,67 @@
+    argnames="value", cases=".", filter=filters.has_tag(B)
+)
+def test_filter_with_tag(value):
+    print(value)
```
```suggestion
    pass
```

smarie commented on pytest_cases/tests/pytest_extension/parametrize_plus/test_filter.py:
```diff
@@ -0,0 +1,67 @@
+    argnames="value", cases=".", filter=~filters.has_tag(B)
+)
+def test_filter_without_tag(value):
+    print(value)
```
```suggestion
    pass
```

smarie commented on pytest_cases/tests/pytest_extension/parametrize_plus/test_filter.py:
```diff
@@ -0,0 +1,67 @@
+    argnames="value", cases=".", filter=filters.has_tag(B) & filters.has_tag(C)
+)
+def test_filter_with_and_relation(value):
+    print(value)
```
```suggestion
    pass
```

smarie commented on pytest_cases/tests/pytest_extension/parametrize_plus/test_filter.py:
```diff
@@ -0,0 +1,67 @@
+    argnames="value", cases=".", filter=filters.has_tag(B) | filters.has_tag(C)
+)
+def test_filter_with_or_relation(value):
+    print(value)
```
```suggestion
    pass
```

smarie commented on pytest_cases/tests/pytest_extension/parametrize_plus/test_filter.py:
```diff
@@ -0,0 +1,67 @@
+    argnames="value", cases=".", filter=filters.id_has_prefix("t")
+)
+def test_filter_with_prefix(value):
+    print(value)
```
```suggestion
    pass
```

smarie commented on pytest_cases/tests/pytest_extension/parametrize_plus/test_filter.py:
```diff
@@ -0,0 +1,67 @@
+    argnames="value", cases=".", filter=filters.id_has_suffix("m")
+)
+def test_filter_with_suffix(value):
+    print(value)
```
```suggestion
    pass


def test_synthesis(module_results_dct):
    assert list(modeul_results_dct) == [
        "test_filter_with_tag[tim]",
        "test_filter_with_tag[tony]",
        "test_filter_without_tag[tom]",
        "test_filter_without_tag[dom]",
        "test_filter_with_and_relation[]",
        "test_filter_with_and_relation[]",
    ]

```

smarie commented on pytest_cases/tests/pytest_extension/parametrize_plus/test_filter.py:
```diff
@@ -0,0 +1,67 @@
+    return 3
+
+
+@case(tags=[A, C], id="dom")
+def case_four():
```
```suggestion
@case(tags=[A, C])
def case_dom():
```

smarie commented:
Thanks for your understanding @saroad2 .

Final round of review: the test should be a meta-test as all others, in other words we should use pytest-harvest to collect the list of test ids in the module and test that this list is correct. The prints can be discarded.

I suggested this in the comments, as you'll see. Can you make sure that it passes ? Indeed I did not try locally, I did it directly in github web page :)

saroad2 commented:
Hey @smarie .

I made the final change and all tests passed locally. The only thing I didn't manage to do is to change `case_four` to `case_dom`. The id of that case was changed and I don't know why.

Other than this, everything looks great :)

smarie commented:
Perfect, thanks @saroad2  ! Merging now
